src/controller/tests/data_model/TestRead.cpp

```diff
@@ -991,6 +1006,153 @@ void TestReadInteraction::TestReadFabricScopedWithFabricFilter(nlTestSuite * apS
+    {
+        mOnError++;
+        mLastError = aError;
+    }
+
+    void OnSubscriptionEstablished(uint64_t aSubscriptionId) override { mOnSubscriptionEstablishedCount++; }
+
+    void ClearCounters()
+    {
+        mAttributeCount                 = 0;
+        mOnReportEnd                    = 0;
+        mOnSubscriptionEstablishedCount = 0;
+        mOnDone                         = 0;
+        mOnError                        = 0;
+        mLastError                      = CHIP_NO_ERROR;
+    }
+
+    uint32_t mAttributeCount                 = 0;
+    uint32_t mOnReportEnd                    = 0;
+    uint32_t mOnSubscriptionEstablishedCount = 0;
+    uint32_t mOnDone                         = 0;
+    uint32_t mOnError                        = 0;
+    CHIP_ERROR mLastError                    = CHIP_NO_ERROR;
+} readCallback;
+
+// void ResubscribePolicy(uint32_t aNumCumulativeRetries, uint32_t & aNextSubscriptionIntervalMsec, bool & aShouldResubscribe)
+// {
+//     aShouldResubscribe = false;
+// }
+} // namespace SubscriptionPathQuotaHelpers
+
+void TestReadInteraction::TestReadHandler_KillOverQuotaSubscriptions(nlTestSuite * apSuite, void * apContext)
+{
+    constexpr size_t kNumberOfInterestedPaths = 10;
+    using namespace SubscriptionPathQuotaHelpers;
+    TestContext & ctx  = *static_cast<TestContext *>(apContext);
+    auto sessionHandle = ctx.GetSessionBobToAlice();
+
+    readCallback.ClearCounters();
+
+    app::InteractionModelEngine::GetInstance()->RegisterReadHandlerAppCallback(&gTestReadInteraction);
+
+    std::vector<std::unique_ptr<app::ReadClient>> readClients;
+
+    app::AttributePathParams attributePaths[kNumberOfInterestedPaths];
+    for (size_t i = 0; i < kNumberOfInterestedPaths; i++)
+    {
+        attributePaths[i] = app::AttributePathParams(kTestEndpointId, TestCluster::Id, TestCluster::Attributes::Int16u::Id);
+    }
+    app::ReadPrepareParams readParams(sessionHandle);
+    readParams.mpAttributePathParamsList    = attributePaths;
+    readParams.mAttributePathParamsListSize = kNumberOfInterestedPaths;
+    readParams.mMaxIntervalCeilingSeconds   = 1;
+    readParams.mKeepSubscriptions           = true;
+
+    // Intentially issue read requests that exceeds the quota per fabric.
+    app::InteractionModelEngine::GetInstance()->SetForceHandlerQuota(false);
+    for (size_t i = 0; i < CHIP_IM_MAX_NUM_READ_HANDLER; i++)
+    {
+        std::unique_ptr<app::ReadClient> readClient =
+            std::make_unique<app::ReadClient>(app::InteractionModelEngine::GetInstance(), &ctx.GetExchangeManager(), readCallback,
+                                              app::ReadClient::InteractionType::Subscribe);
+        NL_TEST_ASSERT(apSuite, readClient->SendRequest(readParams) == CHIP_NO_ERROR);
+        readClients.push_back(std::move(readClient));
+    }
+
+    // There are too many messages and the test (gcc_debug, which includes many sanity checks) will be quite slow. Note: report
```

Comment:
Why do we need to go all the way to `CHIP_IM_MAX_NUM_READ_HANDLER`? Why wouldn't we just artifically scale that down to a more mangeable number for the purposes of testing?

Reply:
Our goal isn't to hit the limits of that define. Our goal is to validate the corners of the algorithm, which we can do with smaller counts.

Reply:
Resolved